topdown/net.go

```diff
@@ -0,0 +1,51 @@
+package topdown
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+type lookupIPAddrCacheKey string
+
+func builtinLookupIPAddr(bctx BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	name, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	key := lookupIPAddrCacheKey(name)
+	if val, ok := bctx.Cache.Get(key); ok {
+		return iter(val.(*ast.Term))
+	}
+
+	addrs, err := net.DefaultResolver.LookupIPAddr(bctx.Context, string(name))
+	if err != nil {
+		var derr *net.DNSError
+		if errors.As(err, &derr) && derr.Temporary() {
```

Comment:
What's the reasoning behind making `derr.Temporary()` halt evaluation? Is it possible to create a test case for this or is it difficult to mock out?

Reply:
I wasn't sure about the analogous cases! If that's so, then we shouldn't halt.

For testing, we'd either follow the path laid out here, https://github.com/foxcpp/go-mockdns/blob/master/resolver.go, or pull it in as a dependency...  shouldn't be to bad if we only implement the one method we need, I'll give that a try 👍

Reply:
OK so stdlib `net` is really not mock friendly, `net.Resolver` is a struct (not interface). What they do in that lib is to spin up a dns server and patch the Resolver struct to override dial, calling their server instead.

Since it's pretty lightweight, I'm going to play with pulling that dependency in. The weird bit is that we'll pull in a dns lib that exceeds net.LookupIP by far, but only need it for testing. Anyhow 🤷 let's see how it goes.

Reply:
Tried it. It's not so bad, if you look the other way on any `golang.org/x/net` dependencies 🙈

Reply:
There's other ways to mock this out, but this has a nice advantage: we get actual stdlib `net` errors, and we should be getting them in the same way if they ever change in the future. Were we to mock the `LookupIPAddr()` call via some interface, we'd have to return hand-crafted errors, and those would not give us many guarantees in the future. 🤔